apps/mail/base.py

```diff
@@ -0,0 +1,89 @@
+from django.contrib import messages
+from django.template import engines
+from django.utils.html import format_html
+from django.utils.text import get_valid_filename
+from django.urls import reverse
+
+from . import tasks
+from .models import Mail
+from .libs.serialize import serialize
+from .libs.camel_space_to_spaces import camel_space_to_spaces
+
+
+class MailBase:
+    @classmethod
+    def process_context(cls, user, request=None, **kwargs):
+        return {
+            'user': user,
+            'request': request,
+            **kwargs
+        }
+
+    @classmethod
+    def get_name(cls):
+        try:
+            return cls.name
+        except AttributeError:
+            name = cls.__name__
+            if name.startswith('Mail'):
+                name = name[4:]
+            return name
+
+    @classmethod
+    def get_subject(cls, user=None, request=None, **kwargs):
+        try:
+            return cls.subject
+        except AttributeError:
+            name = cls.get_name()
+            subject = camel_space_to_spaces(name).capitalize()
+            return subject
+
+    @classmethod
+    def get_template(cls, user=None, request=None, **kwargs):
+        try:
+            return cls.template
+        except AttributeError:
+            name = cls.get_name()
+            path = 'email/{name}.html'.format(
+                name=get_valid_filename(camel_space_to_spaces(name))
+            )
+            return path
+
+    @classmethod
+    def render_body(cls, ctx):
+        django_engine = engines['django']
+        template = django_engine.get_template(cls.get_template())
+        return template.render(ctx)
+
+    @classmethod
+    def render_subject(cls, ctx):
+        django_engine = engines['django']
+        template = django_engine.from_string(cls.get_subject())
+        return template.render(ctx)
+
+    def __new__(cls, user, request=None, **kwargs):
+        ctx = serialize(cls.process_context(user, request, **kwargs))
+        body = cls.render_body(ctx)
+        subject = cls.render_subject(ctx)
+        mail = Mail.objects.create(
+            name=cls.get_name(),
+            user=user,
+            data=ctx,
+            body=body,
+            subject=subject
+        )
+        tasks.send_email(mail.id) # Sets a background task to send the email
+
+        if (request != None):
+            mail_url = reverse(
+                'admin:%s_%s_change' % ('mail', 'mail'),
+                args=[mail.id],
+            )
+            msg = format_html(
+                'The mail message "<a href ="{}">{}</a>" was generated successfully.',
+                mail_url,
+                mail,
+            )
+            messages.success(request, msg)
```

Comment:
How is this utilized? Would this show up in the django admin? Does DRF hook into the messages somehow?

Reply:
Yes, this is purely feedback at the top of the page for Django Admin actions - and a nice way to get a quick look at the mail you just triggered. (I found it pretty handy for development on Floorspace, testing trigger conditions and template updates)

I haven't messages having any effect on DRF (or workers), so I assume they're just discarded in those instances, but I'll look into that.

Might be worth adding a boolean flag to the class, `generate_admin_messages` or something, to disable these?

Reply:
Ah OK, that makes more sense. It's likely fine as is. I doubt DRF pays any attention to these. Boolean could be nice, but I'd say isn't a requirement.